verification/spec/relayer/relayer.tla

```diff
@@ -0,0 +1,225 @@
+------------------------------ MODULE relayer ------------------------------
+
+EXTENDS Naturals, FiniteSets
+
+CONSTANTS MaxHeight, \* maximal height of all the chains in the system 
+          NrChains \* number of chains in the system
+
+VARIABLES chains, \* a function that assigns to each chainID a chain record
+          pendingDatagrams, \* a function that assigns to each chainID a set of pending datagrams
+          relayer, \* a relayer record
+          turn
+
+\* Instance of the environment,  
+\* that takes care of the chain logic    
+Env == INSTANCE environment 
+       WITH chains <- chains, 
+            pendingDatagrams <- pendingDatagrams,
+            MaxHeight <- MaxHeight,
+            NrChains <- NrChains
+                               
+vars == <<chains, pendingDatagrams, relayer, turn>>            
+
+ChainIDs == 1..NrChains
+ClientIDs == 1..(NrChains * NrChains)
+Heights == 1..MaxHeight
+RelayerClientIDs == 1..NrChains
+
+nullClientID == 0
+nullHeight == 0 
+ 
+(***************
+Client ID operators
+***************)
+
+GetChainIDs(clientID) == Env!GetChainIDs(clientID)
+    
+GetClientID(onChainID, forChainID) == Env!GetClientID(onChainID, forChainID)
+    
+ValidClientID(chainID, clientID) == Env!ValidClientID(chainID, clientID) 
+ 
+(***************
+Client datagrams
+****************) 
+ 
+\* get the latest height of the chainID
+GetLatestHeight(chainID) ==
+    chains[chainID].height    
+    
+\* get the client height of the client on chain srcChainID for chain dstChainID    
+GetClientHeight(srcChainID, dstChainID) ==
+    chains[srcChainID].clientHeights[dstChainID]
+
+ClientDatagrams(srcChainID, dstChainID) ==
+    LET dstChainHeight == GetLatestHeight(dstChainID) IN    
+    LET dstClientHeight == GetClientHeight(srcChainID, dstChainID) IN
+    LET srcDatagrams == 
+        IF dstClientHeight = nullHeight
+        THEN \* the dst client does not exist at the src chain 
+            {[type |-> "CreateClient", 
+              height |-> relayer.clientHeights[dstChainID],
+              clientID |-> GetClientID(srcChainID, dstChainID)]} 
+        ELSE \* the dst client exists at the src chain
+            IF dstClientHeight < dstChainHeight
+            THEN \* the height of the dst client is smaller than the height of the dst chain
+                {[type |-> "ClientUpdate",
+                  height |-> relayer.clientHeights[dstChainID],
+                  clientID |-> GetClientID(srcChainID, dstChainID)]} 
+            ELSE {} IN
+                    
+    LET srcChainHeight == GetLatestHeight(srcChainID) IN    
+    LET srcClientHeight == GetClientHeight(dstChainID, srcChainID) IN
```

Comment:
Can we extract this logic into a function? It seems that it is the same logic in these two blocks.